internal/engine/ingester/diff/parse.go

```diff
@@ -161,62 +160,69 @@ func goParse(patch string) ([]*pb.Dependency, error) {
 	return deps, nil
 }
 
-type npmDependency struct {
-	Version      string                    `json:"version"`
-	Resolved     string                    `json:"resolved"`
-	Integrity    string                    `json:"integrity"`
-	Requires     map[string]string         `json:"requires,omitempty"`
-	Dependencies map[string]*npmDependency `json:"dependencies,omitempty"`
-	Optional     bool                      `json:"optional,omitempty"`
-}
-
-type npmRoot struct {
-	Deps map[string]*npmDependency
-}
-
-// TODO(jakub): this is a hacky way to parse the npm patch file
 func npmParse(patch string) ([]*pb.Dependency, error) {
 	lines := strings.Split(patch, "\n")
-	var output strings.Builder
-
-	// Write the start of the JSON object to the output
-	output.WriteString("{\n")
-
-	// Start of crazy code to parse the patch file
-	// What we do here is first grab all the lines that start with "+"
-	// Then we remove the "+" symbol and write the modified line to the output
-	// We then add a { to the start of the output and a } to the end, so we have a valid JSON object
-	// Then we convert the string builder to a string and unmarshal it into the Package struct
-	for _, line := range lines {
-		// Check if the line starts with "+"
-		if strings.HasPrefix(line, "+") {
-			// Remove the "+" symbol and write the modified line to the output
-			line = strings.TrimPrefix(line, "+")
-			output.WriteString(line + "\n")
+	versionRegex := regexp.MustCompile(`^\+\s*"version"\s*:\s*"([^"\n]*)"\s*(?:,|$)`)
+
+	var deps []*pb.Dependency
+
+	for i, line := range lines {
+		// Check if the line contains a version
+		matches := versionRegex.FindStringSubmatch(line)
+		if len(matches) > 1 {
+			version := matches[1]
+			name := findDependencyName(i, lines)
+			// The version is not always a dependency version. It may also be the version of the package in this repo,
+			// or the version of the root project. See https://docs.npmjs.com/cli/v10/configuring-npm/package-lock-json
+			if name != "" {
+				deps = append(deps, &pb.Dependency{
+					Ecosystem: pb.DepEcosystem_DEP_ECOSYSTEM_NPM,
+					Name:      name,
+					Version:   version,
+				})
+			}
 		}
 	}
 
-	//if string ends with a comma, remove it
-	outputString := strings.TrimSuffix(output.String(), ",\n")
-	outputString = outputString + "\n}"
+	return deps, nil
+}
 
-	// Convert the string builder to a string and unmarshal it into the Package struct
-	root := &npmRoot{
-		Deps: make(map[string]*npmDependency),
+// findDependencyName iterates over all the previous lines to find the JSON key containing the parent dependency name.
+// If the parent key does not correspond to a dependency (i.e. it could be the root project), then an empty string is
+// returned.
+func findDependencyName(i int, lines []string) string {
+	dependencyNameRegex := regexp.MustCompile(`\s*"([^"]+)"\s*:\s*{\s*`)
```

Comment:
do you think we could have these as variables global but not exported so we avoid compiling the regex on each call? same for the one in npmParse

Reply:
Makes sense, updating now